packages/did/src/Did.chain.ts

```diff
@@ -105,48 +150,119 @@ function decodeDidChainRecord(
   return didRecord
 }
 
+// Same reasoning as `assembleKeyId`.
+function decodeServiceChainRecord(
+  serviceDetails: IServiceEndpointChainRecordCodec,
+  did: IDidDetails['did']
+): IDidServiceEndpoint {
+  const decodedId = hexToString(serviceDetails.id.toString())
+  return {
+    id: assembleDidFragment(did, decodedId),
+    types: serviceDetails.serviceTypes.map((type) =>
+      hexToString(type.toString())
+    ),
+    urls: serviceDetails.urls.map((url) => hexToString(url.toString())),
+  }
+}
+
 export async function queryById(
   didIdentifier: IIdentity['address']
 ): Promise<IDidChainRecordJSON | null> {
-  const result = await queryEncoded(didIdentifier)
-  if (result.isSome) {
-    return decodeDidChainRecord(
-      result.unwrap(),
-      getKiltDidFromIdentifier(didIdentifier, 'full')
-    )
+  const result = await queryDidEncoded(didIdentifier)
+  if (result.isNone) {
+    return null
   }
-  return null
+  return decodeDidChainRecord(
+    result.unwrap(),
+    getKiltDidFromIdentifier(didIdentifier, 'full')
+  )
 }
 
-export async function queryByDID(
-  did: IDidChainRecordJSON['did']
+// Query full DID details given the DID URI.
+export async function queryDidDetails(
+  didUri: IDidDetails['did']
 ): Promise<IDidChainRecordJSON | null> {
-  // we will have to extract the id part from the did string
-  const didId = getIdentifierFromKiltDid(did)
-  return queryById(didId)
+  const { identifier, fragment } = parseDidUrl(didUri)
+  if (fragment) {
+    throw new Error(`The provided URI ${didUri} must not contain any fragment.`)
+  }
+  return queryById(identifier)
 }
 
-export async function queryKey(
-  did: string,
-  keyId: string
+// Query a given key given the DID identifier and the key ID.
+export async function queryDidKey(
+  keyUri: IDidKeyDetails['id']
 ): Promise<IDidKeyDetails | null> {
-  const encoded = await queryEncoded(getIdentifierFromKiltDid(did))
-  if (encoded.isNone) return null
-  const keyIdU8a = Crypto.coToUInt8(keyId)
-  let key: IDidKeyDetails | null = null
-  encoded.unwrap().publicKeys.forEach((keyDetails, id) => {
-    if (id.eq(keyIdU8a)) {
-      key = decodeDidPublicKeyDetails(did, id, keyDetails)
-    }
-  })
-  return key
+  const { identifier, fragment } = parseDidUrl(keyUri)
+  if (!fragment) {
+    throw new Error(
+      `The provided URI ${keyUri} does not contain a valid fragment for key ID.`
+    )
+  }
+  const didDetails = await queryById(identifier)
+  if (!didDetails) {
+    return null
+  }
+  return (
+    didDetails.publicKeys
+      .filter((key) => {
+        return key.id === keyUri
+      })
+      .pop() || null
```

Comment:
S: You could also just use the `find` method here

Reply:
https://github.com/KILTprotocol/sdk-js/pull/428/commits/4bece8bd3f4c8c70569738559cfd7d078628af36